src/go/k8s/pkg/resources/statefulset.go

```diff
@@ -918,3 +920,48 @@ func (r *StatefulSetResource) CurrentVersion(ctx context.Context) (string, error
 	}
 	return stsVersion, nil
 }
+
+func (r *StatefulSetResource) getPodByBrokerID(ctx context.Context, brokerID *int32) (*corev1.Pod, error) {
+	if brokerID == nil {
+		return nil, nil
+	}
+	brokerIDStr := strconv.FormatInt(int64(*brokerID), 10)
+	pods, err := r.getPodList(ctx)
+	if err != nil {
+		return nil, err
+	}
+	for i := range pods.Items {
+		annotations := pods.Items[i].GetAnnotations()
+		if v, ok := annotations[PodAnnotationNodeIDKey]; ok && v == brokerIDStr {
+			return &pods.Items[i], nil
+		}
+	}
+
+	return nil, nil
+}
+
+func (r *StatefulSetResource) getBrokerIDForPod(ctx context.Context, ordinal int32) (*int32, error) {
```

Comment:
I'm not sure if 
https://github.com/redpanda-data/redpanda/blob/dceb7a892f559d6b8bf98d5c9243e090784d2973/src/go/k8s/controllers/redpanda/cluster_controller.go#L601-L628

should be used here as it is more accurate representation of current Redpanda Node ID. 

The problem with annotation is that given following scenario:

1. Cluster cr created with replicas 5
2. StatefulSet is set to initial replicas 1
3. Scale function handles the scaling from 1 to 5 replicas
4. Cluster did not finish it's Configuration 👇 so annotation is not yet set https://github.com/redpanda-data/redpanda/blob/dceb7a892f559d6b8bf98d5c9243e090784d2973/src/go/k8s/controllers/redpanda/cluster_controller.go#L309-L317
5. Cluster already registered Redpanda Node ID 4 and 3
6. User changes Cluster cr replication to 3
7. Scaling down will break as `getBrokerIDForPod` will return nil as Redpanda Node ID and whole reconciliation will be stuck.

I know it's niche scenario. Most of the time annotation will be set.

Reply:
I kind-of agree but will have to think about it. It might not make this PR due to time constraints but we can come back at it later.

Reply:
I found a way to set the annotation early in the reconcile so this should no longer be an issue.